### ISSUE TITLE
The data decorator does not integrate well with numpydoc

### ISSUE THREAD

**Issue by timhoffm** [matplotlib/matplotlib#10189]
### Bug report

**Bug summary**
From a discussion in #10186: The data note inserted in the docstring via `@_preprocess_data` can be displayed in unexpected places in the generated html documentation.

**Examples**
- in [Other Parmeters](https://matplotlib.org/api/_as_gen/matplotlib.axes.Axes.errorbar.html#matplotlib.axes.Axes.errorbar)
- in [See Also](https://matplotlib.org/api/_as_gen/matplotlib.axes.Axes.vlines.html#matplotlib.axes.Axes.vlines)

**Reason**
I'm not an expert in numpydoc, but from my understanding, this is the issue:

-  `@_preprocess_data`  just appends an admonition
  ~~~
  .. note::
      In addition to the above described arguments, this function can take a
      **data** keyword argument. If such a **data** argument is given, the
  ~~~

- In numpydoc, everything is part of some section. Since  `@_preprocess_data` does not add a section, the admonition belongs to whatever section was last. For sections that are plain top-level such as *Notes* or *Examples*, this looks ok (though technically the note is part of the example). It goes wrong for differently formatted or reordered sections (e.g. indented *Other Parameters* or contained in a *See Also* box). Example docstring:

~~~
See Also
--------
fill_betweenx : Fill between two sets of x-values.

.. note::
    In addition to the above described arguments, this function can take a
    **data** keyword argument. If such a **data** argument is given, the
~~~

**Solution**
To do this correctly, the inserted admonition should always be part of a *Notes* section. There are diffent ways to achieve this:

1. Make sure docstrings using `@_preprocess_data` do always end with a *Notes* section. Manually add an empty one if there is none.
2. Let `@_preprocess_data` add the *Notes* section. This is not trivial because you may or may not have an *Notes* section already and duplicated sections are not allowed (https://github.com/numpy/numpydoc/pull/67 could help on this).
3. Change `@_preprocess_data` to not append someting, but format.  Advantage: This is more explicit and prevents surprises. Disadavantage: The documenter has to know that the note has to be added and do it manually (But at least `@_preprocess_data` could warn if it could not format its note into the docstring). The docstring could look like this:
  ~~~
  Notes
  -----
  {data_note}
  ~~~

There may be other options like extending numpydoc to have an additional custom *Data Note* section, but I don't think its worth the effort.
 
For now, I use option 1. as a simple workaround on a case by case basis (also in #10186).
2. would be too much magic.
In the long run, I propose to switch to option 3.

**PS**
Just got another idea possibly worth considering. If we change to formatting anyway, one could get rid of the admonition and format a custom *data* entry into the *Paramters* or *Additonal Paramters* section:

~~~
Parameters
----------
x : array
    some text
y : array
    other text
{data_param}
color : Color
    more params   
~~~

**Comment by anntzer** [comment]
I vote for the PS option (format a custom data entry).

**Comment by tacaswell** [comment]
The decorator used to add a `Notes` section at the bottom which was the trigger for https://github.com/numpy/numpydoc/pull/65 

Putting the information in Parameters or Other Parameters is best, not sure the right way to do that though.  

Putting in a templatting string in works but requires coordination between two places (the decorator and the docstring).  Having the decorator parse the docstring as numpydoc, insert the lines and then write it back out is also an option, but I am not sure what the start-up time cost of that is. We already have some pretty significant docstring templating going on (and discussions about removing it) so I am not sure that the net run-time cost will be that large.

Another option is a helper that just re-writes the source to extend the docstring, but that is probably the worst of all worlds.

**Comment by timhoffm** [comment]
> Putting the information in Parameters or Other Parameters is best, not sure the right way to do that though.

Why not?

> Putting in a templatting string in works but requires coordination between two places (the decorator and the docstring).

That's true, but the decorator can easily check if it could modify the docstring using formatting and warn otherwise.

**Comment by timhoffm** [comment]
Until this is resolved, I will use a standard empty *Notes* section

        Notes
        -----
        .. [Notes section required for data comment. See #10189.]

This will make a possible later migration easier.

**Comment by dstansby** [comment]
Looks like this will be fixed by #14029

**Comment by timhoffm** [comment]
Fixed by  #14029.
